app/Relations/HasAlbumThumb.php

```diff
@@ -0,0 +1,241 @@
+	 * because the DBMS can use its indexes.
+	 *
+	 * For PostgreSQL we could use the `DISTINCT ON`-clause to achieve the
+	 * result:
+	 *
+	 *     SELECT DISTINCT ON (covered_album_id)
+	 *       covered_albums.id AS covered_album_id,
+	 *       photos.id         AS id,
+	 *       photos.type       AS type
+	 *     FROM covered_albums
+	 *     LEFT JOIN
+	 *       (
+	 *         photos
+	 *         LEFT JOIN albums
+	 *         ON (albums.id = photos.album_id)
+	 *       )
+	 *     ON (
+	 *       albums._lft >= covered_albums._lft AND
+	 *       albums._rgt <= covered_albums._rgt AND
+	 *       "complicated seachability filter goes here"
+	 *     )
+	 *     WHERE covered_albums.id IN $albumKeys
+	 *     ORDER BY album_id ASC, photos.is_starred DESC, photos.created_at DESC
+	 *
+	 * For PostgreSQL see ["SELECT - DISTINCT Clause"](https://www.postgresql.org/docs/13/sql-select.html#SQL-DISTINCT).
+	 *
+	 * But `DISTINCT ON` is neither provided by MySQL and SQLite.
+	 * For the latter two, the following non-SQL-conformant query could be
+	 * used:
+	 *
+	 *     SELECT
+	 *       covered_albums.id  AS covered_album_id,
+	 *       photos.id          AS id,
+	 *       photos.type        AS type
+	 *     FROM covered_albums
+	 *     LEFT JOIN
+	 *       (
+	 *         photos
+	 *         LEFT JOIN albums
+	 *         ON (albums.id = photos.album_id)
+	 *       )
+	 *     ON (
+	 *       albums._lft >= covered_albums._lft AND
+	 *       albums._rgt <= covered_albums._rgt AND
+	 *       "complicated seachability filter goes here"
+	 *     )
+	 *     WHERE covered_albums.id IN $albumKeys
+	 *     ORDER BY album_id ASC, photos.is_starred DESC, photos.created_at DESC
+	 *     GROUP BY album_id
+	 *
+	 * Instead of enforcing distinct results for `covered_album_id`, the result
+	 * is grouped by `covered_album_id`.
+	 * Note that this is not SQL-compliant, because the `SELECT` clause
+	 * contains two columns (`photo.id` and `photo.type`) which are neither
+	 * part of the `GROUP BY`-clause nor aggregates.
+	 * However, MySQL and SQLite relax this constraint and return the
+	 * column values of the first row of a group.
+	 * This is exactly the specified behaviour of `DISTINCT ON`.
+	 * For SQLite see "[Quirks, Caveats, and Gotchas In SQLite, Sec. 6](https://www.sqlite.org/quirks.html)"
+	 *
+	 * TODO: If the following query is too slow for large installation, we must write two separate implementations for PostgreSQL and MySQL/SQLite as outlined above.
+	 *
+	 * @param array<Album> $models
+	 */
+	public function addEagerConstraints(array $models): void
+	{
+		$albumKeys = $this->getKeys($models);
+
+		$bestPhotoIDSelect = Photo::query()
+			->select(['photos.id AS photo_id'])
+			->leftJoin('albums', 'albums.id', '=', 'photos.album_id')
+			->whereColumn('albums._lft', '>=', 'covered_albums._lft')
+			->whereColumn('albums._rgt', '<=', 'covered_albums._rgt')
+			->orderBy('photos.is_starred', 'desc')
+			->orderBy('photos.created_at', 'desc')
```

Comment:
Right. I wonder where I got this from. However, there is one aspect which we need to discuss.

If the thumbnails of albums are determined iteratively, i.e. with one SQL query per album, then each SQL query can use the sorting order which is specifically set for the album at hand. If the thumbnails for all albums are fetched eagerly, then all photos which might be picked as a thumbnail of all albums must be prioritized according to the same criterion. Of course, this criterion does not need to be `created_at` it can still be the default sorting criterion from the settings.

But either way that may lead to inconsistencies which photo is selected as a thumbnail depending on wether the thumbnails are fetched individually or eagerly.

So I wonder if we should always prioritize the images according to the default ordering.

Reply:
As said in the gitter, currently the thumbs are determined by the ordering of the album (this has been the case since v3). But I am open to changes for faster approaches. Do note that now we can specify the thumb of the album so a user now has the ability to select which picture to use.

Reply:
Yeah, I think it should at the least use the global sorting order set in the Config.

So can you explain when exactly it is using eager vs individual?

Looking at the code I'm not sure if the eager variant actually respects the explicit choice of the cover thumb specified by the user? If not, then that would be _really_ hard to swallow...

Reply:
> Looking at the code I'm not sure if the eager variant actually respects the explicit choice of the cover thumb specified by the user? If not, then that would be really hard to swallow...

Right, I know that this is still missing, but I am aware of that regression and also have an idea how to fix it. I just haven't found the time to code this.

Reply:
The problem with the missing special handling of explicitly set covers is fixed now.